File: golang/3147.taking-maximum-energy-from-the-mystic-dungeon.go
func maximumEnergy(energy []int, k int) int {
	n := len(energy)
	prefix := make([]int, n)

	for i := range n {
		prefix[i] = energy[i]

		if i >= k {
			prefix[i] = max(prefix[i], prefix[i]+prefix[i-k])
		}
	}

	from := max(0, n-k)

	return slices.Max(prefix[from:])
}

File: python/3147.taking-maximum-energy-from-the-mystic-dungeon.py
class Solution:
    def maximumEnergy(self, energy: list[int], k: int) -> int:
        n = len(energy)
        prefix = [0] * n

        for i in range(n):
            prefix[i] = energy[i]

            if i - k >= 0:
                prefix[i] = max(prefix[i], prefix[i] + prefix[i - k])

        return max(prefix[-k:])

